File: src/api/musicSource/library.js
/**
 * 音乐库管理 API — 后端 library_* 命令
 *
 * 所有返回值通过 {@link Song|Artist|Album|Lyric} 类包装。
 */

import { transport } from '@/api/transport';
import { Song, Artist, Album, Lyric, ArtistSummary, AlbumSummary } from '@/class';

// ══════════════════════════════════════════════════════════════════════════════
// Persistence
// ══════════════════════════════════════════════════════════════════════════════

/** 立即将所有未落盘的修改写入磁盘 */
export async function save() {
  return transport.command('library_save');
}

/** 清理所有 source_ids 为空的实体 */
export async function cleanupEmptyEntities() {
  return transport.command('library_cleanup_empty_entities');
}

// ══════════════════════════════════════════════════════════════════════════════
// Song
// ══════════════════════════════════════════════════════════════════════════════

/** @returns {Promise<number>} */
export async function songCount() {
  return transport.command('library_song_count');
}

/** @param {string} id @returns {Promise<Song>} */
export async function getSong(id) {
  const data = await transport.command('library_get_song', { id });
  return new Song(data);
}

/** @returns {Promise<Record<string, Song>>} */
export async function getAllSongs() {
  const data = await transport.command('library_get_all_songs');
  const map = {};
  for (const [id, d] of Object.entries(data)) {
    map[id] = new Song(d);
  }
  return map;
}

/**
 * 分页获取歌曲，减少 IPC 载荷。
 * @param {number} offset
 * @param {number} limit
 * @returns {Promise<{songs: Song[], total: number}>}
 */
export async function getSongsPage(offset = 0, limit = 50) {
  const [songsData, total] = await Promise.all([
    transport.command('library_get_songs_page', { offset, limit }),
    transport.command('library_song_count'),
  ]);
  return {
    songs: Song.fromDataArray(songsData),
    total,
  };
}

/** @param {string} query @returns {Promise<Song[]>} */
export async function searchSongs(query) {
  const data = await transport.command('library_search_songs', { query });
  return Song.fromDataArray(data);
}

// ══════════════════════════════════════════════════════════════════════════════
// Artist
// ══════════════════════════════════════════════════════════════════════════════

/** @returns {Promise<number>} */
export async function artistCount() {
  return transport.command('library_artist_count');
}

/** @param {string} id @returns {Promise<Artist>} */
export async function getArtist(id) {
  const data = await transport.command('library_get_artist', { id });
  return new Artist(data);
}

/** @returns {Promise<Record<string, Artist>>} */
export async function getAllArtists() {
  const data = await transport.command('library_get_all_artists');
  const map = {};
  for (const [id, d] of Object.entries(data)) {
    map[id] = new Artist(d);
  }
  return map;
}

/**
 * 分页获取艺术家。
 * @param {number} offset
 * @param {number} limit
 * @returns {Promise<{artists: Artist[], total: number}>}
 */
export async function getArtistsPage(offset = 0, limit = 50) {
  const [artistsData, total] = await Promise.all([
    transport.command('library_get_artists_page', { offset, limit }),
    transport.command('library_artist_count'),
  ]);
  return {
    artists: Artist.fromDataArray(artistsData),
    total,
  };
}

/** @param {string} query @returns {Promise<Artist[]>} */
export async function searchArtists(query) {
  const data = await transport.command('library_search_artists', { query });
  return Artist.fromDataArray(data);
}

// ══════════════════════════════════════════════════════════════════════════════
// Album
// ══════════════════════════════════════════════════════════════════════════════

/** @returns {Promise<number>} */
export async function albumCount() {
  return transport.command('library_album_count');
}

/** @param {string} id @returns {Promise<Album>} */
export async function getAlbum(id) {
  const data = await transport.command('library_get_album', { id });
  return new Album(data);
}

/** @returns {Promise<Record<string, Album>>} */
export async function getAllAlbums() {
  const data = await transport.command('library_get_all_albums');
  const map = {};
  for (const [id, d] of Object.entries(data)) {
    map[id] = new Album(d);
  }
  return map;
}

/**
 * 分页获取专辑。
 * @param {number} offset
 * @param {number} limit
 * @returns {Promise<{albums: Album[], total: number}>}
 */
export async function getAlbumsPage(offset = 0, limit = 50) {
  const [albumsData, total] = await Promise.all([
    transport.command('library_get_albums_page', { offset, limit }),
    transport.command('library_album_count'),
  ]);
  return {
    albums: Album.fromDataArray(albumsData),
    total,
  };
}

// ── homeStats 缓存 ────────────────────────────────────────────────────────────
let _homeStatsCache = null;
let _homeStatsCacheTime = 0;
const HOME_STATS_CACHE_TTL = 30_000; // 30 秒内复用

/**
 * 获取首页所需数据 — 一次轻量 IPC 调用替代三次全量加载。
 * 内置 30 秒 TTL 缓存，避免每次导航到首页都重新请求。
 * @param {boolean} [force=false] 设为 true 强制跳过缓存
 * @returns {Promise<{stats: {tracks: number, artists: number, albums: number}, recentTracks: Song[], featuredArtists: Artist[], recentAlbums: Album[]}>}
 */
export async function homeStats(force = false) {
  const now = Date.now();
  if (!force && _homeStatsCache && (now - _homeStatsCacheTime) < HOME_STATS_CACHE_TTL) {
    return _homeStatsCache;
  }
  const data = await transport.command('library_get_home_stats');
  _homeStatsCache = {
    stats: data.stats,
    recentTracks: Song.fromDataArray(data.recentTracks || []),
    featuredArtists: Artist.fromDataArray(data.featuredArtists || []),
    recentAlbums: Album.fromDataArray(data.recentAlbums || []),
  };
  _homeStatsCacheTime = now;
  return _homeStatsCache;
}

/** @param {string} query @returns {Promise<Album[]>} */
export async function searchAlbums(query) {
  const data = await transport.command('library_search_albums', { query });
  return Album.fromDataArray(data);
}

// ══════════════════════════════════════════════════════════════════════════════
// 统一搜索 — 基于 Rust trigram 倒排索引的跨类型快速子串搜索
// ══════════════════════════════════════════════════════════════════════════════

/**
 * 统一搜索 — 同时搜索歌曲、艺术家、专辑，支持按类型与来源过滤。
 *
 * 后端基于 Rust trigram 倒排索引，10k 条目查询约 0.3ms。
 * 索引首次查询时构建并缓存，写操作自动失效。
 *
 * @param {object} opts
 * @param {string} opts.query - 搜索关键词（大小写不敏感子串匹配）
 * @param {'song'|'artist'|'album'|null} [opts.entityType=null] - 限定实体类型，null 全搜
 * @param {string|null} [opts.sourceName=null] - 限定来源名称，null 不限制
 * @param {number|null} [opts.limitPerType=null] - 每类实体最多返回多少条
 * @returns {Promise<{songs: Song[], artists: Artist[], albums: Album[]}>}
 */
export async function search({ query, entityType = null, sourceName = null, limitPerType = null }) {
  const args = { query };
  if (entityType) args.entity_type = entityType;
  if (sourceName) args.source_name = sourceName;
  if (limitPerType != null) args.limit_per_type = limitPerType;
  const data = await transport.command('library_search', args);
  return {
    songs: Song.fromDataArray(data.songs || []),
    artists: Artist.fromDataArray(data.artists || []),
    albums: Album.fromDataArray(data.albums || []),
  };
}

// ══════════════════════════════════════════════════════════════════════════════
// Lyric
// ══════════════════════════════════════════════════════════════════════════════

/** @returns {Promise<number>} */
export async function lyricCount() {
  return transport.command('library_lyric_count');
}

/** @param {string} id @returns {Promise<Lyric>} */
export async function getLyric(id) {
  const data = await transport.command('library_get_lyric', { id });
  return new Lyric(data);
}

/** @returns {Promise<Record<string, Lyric>>} */
export async function getAllLyrics() {
  const data = await transport.command('library_get_all_lyrics');
  const map = {};
  for (const [id, d] of Object.entries(data)) {
    map[id] = new Lyric(d);
  }
  return map;
}

/** @param {string} query @returns {Promise<Lyric[]>} */
export async function searchLyrics(query) {
  const data = await transport.command('library_search_lyrics', { query });
  return Lyric.fromDataArray(data);
}

// ══════════════════════════════════════════════════════════════════════════════
// Relations
// ══════════════════════════════════════════════════════════════════════════════

/** @param {string} songId @returns {Promise<Artist[]>} */
export async function getArtistsOfSong(songId) {
  const data = await transport.command('library_get_artists_of_song', { songId });
  return Artist.fromDataArray(data);
}

/** @param {string} songId @returns {Promise<Album|null>} */
export async function getAlbumOfSong(songId) {
  const data = await transport.command('library_get_album_of_song', { songId });
  return data ? new Album(data) : null;
}

/** @param {string} songId @returns {Promise<Lyric|null>} */
export async function getLyricOfSong(songId) {
  const data = await transport.command('library_get_lyric_of_song', { songId });
  return data ? new Lyric(data) : null;
}

/** @param {string} artistId @returns {Promise<Song[]>} */
export async function getSongsByArtist(artistId) {
  const data = await transport.command('library_get_songs_by_artist', { artistId });
  return Song.fromDataArray(data);
}

/** @param {string} artistId @returns {Promise<Album[]>} */
export async function getAlbumsByArtist(artistId) {
  const data = await transport.command('library_get_albums_by_artist', { artistId });
  return Album.fromDataArray(data);
}

/** @param {string} albumId @returns {Promise<Song[]>} */
export async function getSongsInAlbum(albumId) {
  const data = await transport.command('library_get_songs_in_album', { albumId });
  return Song.fromDataArray(data);
}

/** @param {string} songId @returns {Promise<SourceId[]>} */
export async function getSourceIdsOfSong(songId) {
  const data = await transport.command('library_get_source_ids_of_song', { songId });
  const { SourceId } = await import('@/class');
  return (data || []).map((s) => new SourceId(s));
}

// ══════════════════════════════════════════════════════════════════════════════
// Memory cache — avoids re-fetching the entire library on every navigation
// ══════════════════════════════════════════════════════════════════════════════

let _cache = null;
let _cacheTime = 0;
const CACHE_TTL = 30_000; // 30 秒内重复导航不重新加载

/**
 * 使缓存失效。在数据变更（添加/删除歌曲、重新扫描等）后调用。
 */
export function invalidateCache() {
  _cache = null;
  _cacheTime = 0;
  _homeStatsCache = null;
  _homeStatsCacheTime = 0;
}

// ══════════════════════════════════════════════════════════════════════════════
// Backward compat aliases
// ══════════════════════════════════════════════════════════════════════════════

/** @deprecated use {@link searchSongs} + {@link getAllArtists} + {@link getAllAlbums} */
export async function scanAll() {
  const [songsData, artistsData, albumsData] = await Promise.all([
    transport.command('library_get_all_songs'),
    transport.command('library_get_all_artists'),
    transport.command('library_get_all_albums'),
  ]);
  const songs = Object.values(songsData).map((d) => new Song(d));
  const artists = Object.values(artistsData).map((d) => new Artist(d));
  const albums = Object.values(albumsData).map((d) => new Album(d));
  return { sources: [], tracks: songs, artists, albums };
}

/**
 * 获取缓存的库数据。
 *
 * 30 秒内重复调用会直接返回缓存结果，避免每次页面切换都全量
 * 加载 921 首歌曲 + 全部艺术家 + 全部专辑。
 *
 * @param {boolean} [force=false] 设为 true 强制跳过缓存
 * @deprecated use {@link searchSongs} with source filter
 */
export async function getCached(force = false) {
  const now = Date.now();
  if (!force && _cache && (now - _cacheTime) < CACHE_TTL) {
    return _cache;
  }
  _cache = await scanAll();
  _cacheTime = now;
  return _cache;
}

/** @deprecated use {@link rescanAll} from sources.js */
export async function refreshSource(_sourceId) {
  const { rescanAll } = await import('./sources.js');
  const result = await rescanAll();
  return Song.fromDataArray([]);
}

export const library = {
  save,
  cleanupEmptyEntities,
  songCount,
  getSong,
  getAllSongs,
  getSongsPage,
  searchSongs,
  artistCount,
  getArtist,
  getAllArtists,
  getArtistsPage,
  searchArtists,
  albumCount,
  getAlbum,
  getAllAlbums,
  getAlbumsPage,
  searchAlbums,
  search,
  homeStats,
  lyricCount,
  getLyric,
  getAllLyrics,
  searchLyrics,
  getArtistsOfSong,
  getAlbumOfSong,
  getLyricOfSong,
  getSongsByArtist,
  getAlbumsByArtist,
  getSongsInAlbum,
  getSourceIdsOfSong,
  invalidateCache,
  // deprecated
  scanAll,
  getCached,
  refreshSource,
  getSourceFromCache: refreshSource,
};

export default library;

File: src-tauri/crates/chordial-server/src/routes/library.rs
//! 音乐库路由 — Song / Artist / Album / Lyric CRUD + 搜索 + 关系查询。
//!
//! # CRUD
//! | 方法 | 路径 | 对应命令 |
//! |------|------|---------|
//! | GET | `/library/songs` | `library_get_all_songs` |
//! | GET | `/library/songs/:id` | `library_get_song` |
//! | GET | `/library/songs/search?q=` | `library_search_songs` |
//! | GET | `/library/songs/count` | `library_song_count` |
//! | GET | `/library/artists` | `library_get_all_artists` |
//! | GET | `/library/artists/:id` | `library_get_artist` |
//! | GET | `/library/artists/search?q=` | `library_search_artists` |
//! | GET | `/library/artists/count` | `library_artist_count` |
//! | GET | `/library/albums` | `library_get_all_albums` |
//! | GET | `/library/albums/:id` | `library_get_album` |
//! | GET | `/library/albums/search?q=` | `library_search_albums` |
//! | GET | `/library/albums/count` | `library_album_count` |
//! | GET | `/library/lyrics` | `library_get_all_lyrics` |
//! | GET | `/library/lyrics/:id` | `library_get_lyric` |
//! | GET | `/library/lyrics/search?q=` | `library_search_lyrics` |
//! | GET | `/library/lyrics/count` | `library_lyric_count` |
//! | POST | `/library/save` | `library_save` |
//! | POST | `/library/cleanup` | `library_cleanup_empty_entities` |
//!
//! # Relations
//! | 方法 | 路径 | 对应命令 |
//! |------|------|---------|
//! | GET | `/library/songs/:id/artists` | `library_get_artists_of_song` |
//! | GET | `/library/songs/:id/album` | `library_get_album_of_song` |
//! | GET | `/library/songs/:id/lyric` | `library_get_lyric_of_song` |
//! | GET | `/library/songs/:id/source-ids` | `library_get_source_ids_of_song` |
//! | GET | `/library/artists/:id/songs` | `library_get_songs_by_artist` |
//! | GET | `/library/artists/:id/albums` | `library_get_albums_by_artist` |
//! | GET | `/library/albums/:id/songs` | `library_get_songs_in_album` |

use crate::state::AppState;
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::routing::{get, post};
use axum::{Json, Router};
use serde::Deserialize;

pub fn router() -> Router<AppState> {
    Router::new()
        // 持久化
        .route("/library/save", post(library_save))
        .route("/library/cleanup", post(library_cleanup))
        // Song
        .route("/library/songs", get(get_all_songs))
        .route("/library/songs/count", get(song_count))
        .route("/library/songs/search", get(search_songs))
        .route(
            "/library/songs/:id",
            get(get_song),
        )
        .route("/library/songs/:id/artists", get(artists_of_song))
        .route("/library/songs/:id/album", get(album_of_song))
        .route("/library/songs/:id/lyric", get(lyric_of_song))
        .route("/library/songs/:id/source-ids", get(source_ids_of_song))
        // Artist
        .route("/library/artists", get(get_all_artists))
        .route("/library/artists/count", get(artist_count))
        .route("/library/artists/search", get(search_artists))
        .route(
            "/library/artists/:id",
            get(get_artist),
        )
        .route("/library/artists/:id/songs", get(songs_by_artist))
        .route("/library/artists/:id/albums", get(albums_by_artist))
        // Album
        .route("/library/albums", get(get_all_albums))
        .route("/library/albums/count", get(album_count))
        .route("/library/albums/search", get(search_albums))
        .route(
            "/library/albums/:id",
            get(get_album),
        )
        .route("/library/albums/:id/songs", get(songs_in_album))
        // Lyric
        .route("/library/lyrics", get(get_all_lyrics))
        .route("/library/lyrics/count", get(lyric_count))
        .route("/library/lyrics/search", get(search_lyrics))
        .route(
            "/library/lyrics/:id",
            get(get_lyric),
        )
}

// ── 持久化 ──────────────────────────────────────────

async fn library_save(State(state): State<AppState>) -> Result<StatusCode, String> {
    state.ctx.library.save()?;
    Ok(StatusCode::NO_CONTENT)
}

async fn library_cleanup(State(state): State<AppState>) -> Result<StatusCode, String> {
    state.ctx.library.cleanup_empty_entities()?;
    state.ctx.library.save()?;
    Ok(StatusCode::NO_CONTENT)
}

// ── Song ────────────────────────────────────────────

async fn song_count(State(state): State<AppState>) -> Json<usize> {
    Json(state.ctx.library.song_count())
}

async fn get_song(State(state): State<AppState>, Path(id): Path<String>) -> impl IntoResponse {
    match state.ctx.library.get_song(&id) {
        Some(song) => Ok(Json(serde_json::to_value(&song).unwrap())),
        None => Err((StatusCode::NOT_FOUND, format!("歌曲 '{}' 不存在", id))),
    }
}

async fn get_all_songs(State(state): State<AppState>) -> Json<serde_json::Value> {
    let songs: Vec<_> = state.ctx.library.get_all_songs().into_values().collect();
    Json(serde_json::to_value(&songs).unwrap())
}

#[derive(Debug, Deserialize)]
struct SearchQuery {
    q: String,
}

async fn search_songs(
    State(state): State<AppState>,
    Query(query): Query<SearchQuery>,
) -> Json<serde_json::Value> {
    let songs = state.ctx.library.search_songs(&query.q);
    Json(serde_json::to_value(&songs).unwrap())
}

// ── Artist ──────────────────────────────────────────

async fn artist_count(State(state): State<AppState>) -> Json<usize> {
    Json(state.ctx.library.artist_count())
}

async fn get_artist(State(state): State<AppState>, Path(id): Path<String>) -> impl IntoResponse {
    match state.ctx.library.get_artist(&id) {
        Some(artist) => Ok(Json(serde_json::to_value(&artist).unwrap())),
        None => Err((StatusCode::NOT_FOUND, format!("艺术家 '{}' 不存在", id))),
    }
}

async fn get_all_artists(State(state): State<AppState>) -> Json<serde_json::Value> {
    let artists: Vec<_> = state.ctx.library.get_all_artists().into_values().collect();
    Json(serde_json::to_value(&artists).unwrap())
}

async fn search_artists(
    State(state): State<AppState>,
    Query(query): Query<SearchQuery>,
) -> Json<serde_json::Value> {
    let artists = state.ctx.library.search_artists(&query.q);
    Json(serde_json::to_value(&artists).unwrap())
}

// ── Album ───────────────────────────────────────────

async fn album_count(State(state): State<AppState>) -> Json<usize> {
    Json(state.ctx.library.album_count())
}

async fn get_album(State(state): State<AppState>, Path(id): Path<String>) -> impl IntoResponse {
    match state.ctx.library.get_album(&id) {
        Some(album) => Ok(Json(serde_json::to_value(&album).unwrap())),
        None => Err((StatusCode::NOT_FOUND, format!("专辑 '{}' 不存在", id))),
    }
}

async fn get_all_albums(State(state): State<AppState>) -> Json<serde_json::Value> {
    let albums: Vec<_> = state.ctx.library.get_all_albums().into_values().collect();
    Json(serde_json::to_value(&albums).unwrap())
}

async fn search_albums(
    State(state): State<AppState>,
    Query(query): Query<SearchQuery>,
) -> Json<serde_json::Value> {
    let albums = state.ctx.library.search_albums(&query.q);
    Json(serde_json::to_value(&albums).unwrap())
}

// ── Lyric ───────────────────────────────────────────

async fn lyric_count(State(state): State<AppState>) -> Json<usize> {
    Json(state.ctx.library.lyric_count())
}

async fn get_lyric(State(state): State<AppState>, Path(id): Path<String>) -> impl IntoResponse {
    match state.ctx.library.get_lyric(&id) {
        Some(lyric) => Ok(Json(serde_json::to_value(&lyric).unwrap())),
        None => Err((StatusCode::NOT_FOUND, format!("歌词 '{}' 不存在", id))),
    }
}

async fn get_all_lyrics(State(state): State<AppState>) -> Json<serde_json::Value> {
    let lyrics: Vec<_> = state.ctx.library.get_all_lyrics().into_values().collect();
    Json(serde_json::to_value(&lyrics).unwrap())
}

async fn search_lyrics(
    State(state): State<AppState>,
    Query(query): Query<SearchQuery>,
) -> Json<serde_json::Value> {
    let lyrics = state.ctx.library.search_lyrics(&query.q);
    Json(serde_json::to_value(&lyrics).unwrap())
}

// ── Relations ───────────────────────────────────────

async fn artists_of_song(
    State(state): State<AppState>,
    Path(song_id): Path<String>,
) -> Json<serde_json::Value> {
    let artists = state.ctx.library.get_artists_of_song(&song_id);
    Json(serde_json::to_value(&artists).unwrap())
}

async fn album_of_song(
    State(state): State<AppState>,
    Path(song_id): Path<String>,
) -> impl IntoResponse {
    match state.ctx.library.get_album_of_song(&song_id) {
        Some(album) => Ok(Json(serde_json::to_value(&album).unwrap())),
        None => Err((StatusCode::NOT_FOUND, format!("歌曲 '{}' 没有关联专辑", song_id))),
    }
}

async fn lyric_of_song(
    State(state): State<AppState>,
    Path(song_id): Path<String>,
) -> impl IntoResponse {
    match state.ctx.library.get_lyric_of_song(&song_id) {
        Some(lyric) => Ok(Json(serde_json::to_value(&lyric).unwrap())),
        None => Err((StatusCode::NOT_FOUND, format!("歌曲 '{}' 没有关联歌词", song_id))),
    }
}

async fn source_ids_of_song(
    State(state): State<AppState>,
    Path(song_id): Path<String>,
) -> Json<serde_json::Value> {
    let ids = state.ctx.library.get_source_ids_of_song(&song_id);
    Json(serde_json::to_value(&ids).unwrap())
}

async fn songs_by_artist(
    State(state): State<AppState>,
    Path(artist_id): Path<String>,
) -> Json<serde_json::Value> {
    let songs = state.ctx.library.get_songs_by_artist(&artist_id);
    Json(serde_json::to_value(&songs).unwrap())
}

async fn albums_by_artist(
    State(state): State<AppState>,
    Path(artist_id): Path<String>,
) -> Json<serde_json::Value> {
    let albums = state.ctx.library.get_albums_by_artist(&artist_id);
    Json(serde_json::to_value(&albums).unwrap())
}

async fn songs_in_album(
    State(state): State<AppState>,
    Path(album_id): Path<String>,
) -> Json<serde_json::Value> {
    let songs = state.ctx.library.get_songs_in_album(&album_id);
    Json(serde_json::to_value(&songs).unwrap())
}
